django/template/defaulttags.py

```diff
@@ -1564,3 +1564,172 @@ def do_with(parser, token):
     nodelist = parser.parse(("endwith",))
     parser.delete_first_token()
     return WithNode(None, None, nodelist, extra_context=extra_context)
+
+
+class TemplateProxy:
+    """
+    Wraps nodelist as partial, in order to bind context.
+    """
+
+    def __init__(self, nodelist, origin, name):
+        self.nodelist = nodelist
+        self.origin = origin
+        self.name = name
+
+    def get_exception_info(self, exception, token):
+        template = self.origin.loader.get_template(self.origin.template_name)
+        return template.get_exception_info(exception, token)
+
+    @property
+    def source(self):
+        template = self.origin.loader.get_template(self.origin.template_name)
+        return template.source
+
+    def _render(self, context):
+        return self.nodelist.render(context)
+
+    def render(self, context):
+        "Display stage -- can be called many times"
+        with context.render_context.push_state(self):
+            if context.template is None:
+                with context.bind_template(self):
+                    context.template_name = self.name
+                    return self._render(context)
+            else:
+                return self._render(context)
+
+
+class DefinePartialNode(Node):
+    def __init__(self, partial_name, inline, nodelist):
+        self.partial_name = partial_name
+        self.inline = inline
+        self.nodelist = nodelist
+
+    def render(self, context):
+        """Set content into context and return empty string"""
+        if self.inline:
+            return self.nodelist.render(context)
+        else:
+            return ""
+
+
+class RenderPartialNode(Node):
+    def __init__(self, partial_name, partial_mapping):
+        # Defer lookup of nodelist to runtime.
+        self.partial_name = partial_name
+        self.partial_mapping = partial_mapping
+
+    def render(self, context):
+        return self.partial_mapping[self.partial_name].render(context)
+
+
+@register.tag(name="partialdef")
+def partialdef_func(parser, token):
+    """
+    Declare a partial that can be used later in the template.
+
+    Usage::
+
+        {% partialdef partial_name %}
+        Partial content goes here
+        {% endpartialdef %}
+
+    Stores the nodelist in the context under the key "partial_contents" and can
+    be retrieved using the {% partial %} tag.
+
+    The optional ``inline`` argument will render the contents of the partial
+    where it is defined.
+    """
+    return _define_partial(parser, token, "endpartialdef")
```

Comment:
You should inline this. `_define_partial` only exists to allow the rename of the tags in the original package.